File: ui/goose2/src/shared/types/messages.ts
// Message roles
export type MessageRole = "user" | "assistant" | "system";

// Content block types
export interface TextContent {
  type: "text";
  text: string;
}

export interface ImageContent {
  type: "image";
  source:
    | { type: "base64"; mediaType: string; data: string }
    | { type: "url"; url: string };
}

export type ToolCallStatus =
  | "pending"
  | "executing"
  | "completed"
  | "error"
  | "stopped";

export type MessageCompletionStatus =
  | "inProgress"
  | "completed"
  | "error"
  | "stopped";

export interface ToolRequestContent {
  type: "toolRequest";
  id: string;
  name: string;
  arguments: Record<string, unknown>;
  status: ToolCallStatus;
}

export interface ToolResponseContent {
  type: "toolResponse";
  id: string;
  name: string;
  result: string;
  isError: boolean;
}

export interface ThinkingContent {
  type: "thinking";
  text: string;
}

export interface RedactedThinkingContent {
  type: "redactedThinking";
}

export interface ReasoningContent {
  type: "reasoning";
  text: string;
}

export interface ActionRequiredContent {
  type: "actionRequired";
  id: string;
  actionType: "toolConfirmation" | "elicitation";
  message?: string;
  toolName?: string;
  arguments?: Record<string, unknown>;
  schema?: Record<string, unknown>;
}

export interface SystemNotificationContent {
  type: "systemNotification";
  notificationType: "compaction" | "info" | "warning" | "error";
  text: string;
}

export type MessageContent =
  | TextContent
  | ImageContent
  | ToolRequestContent
  | ToolResponseContent
  | ThinkingContent
  | RedactedThinkingContent
  | ReasoningContent
  | ActionRequiredContent
  | SystemNotificationContent;

export interface MessageAttachment {
  type: "file" | "url" | "directory";
  name: string;
  path?: string;
  url?: string;
}

export interface MessageChip {
  label: string;
  type: "skill" | "extension" | "recipe";
}

export interface MessageMetadata {
  userVisible?: boolean;
  agentVisible?: boolean;
  attachments?: MessageAttachment[];
  chips?: MessageChip[];
  /** Persona that generated this assistant message (set on send). */
  personaId?: string;
  personaName?: string;
  /** Which persona this user message is addressed to. */
  targetPersonaId?: string;
  targetPersonaName?: string;
  completionStatus?: MessageCompletionStatus;
}

export interface Message {
  id: string;
  role: MessageRole;
  created: number;
  content: MessageContent[];
  metadata?: MessageMetadata;
}

// Type guards for content blocks
export function isTextContent(c: MessageContent): c is TextContent {
  return c.type === "text";
}
export function isToolRequest(c: MessageContent): c is ToolRequestContent {
  return c.type === "toolRequest";
}
export function isToolResponse(c: MessageContent): c is ToolResponseContent {
  return c.type === "toolResponse";
}
export function isThinking(c: MessageContent): c is ThinkingContent {
  return c.type === "thinking";
}
export function isReasoning(c: MessageContent): c is ReasoningContent {
  return c.type === "reasoning";
}
export function isActionRequired(
  c: MessageContent,
): c is ActionRequiredContent {
  return c.type === "actionRequired";
}
export function isSystemNotification(
  c: MessageContent,
): c is SystemNotificationContent {
  return c.type === "systemNotification";
}

// Helpers
export function getTextContent(message: Message): string {
  return message.content
    .filter(isTextContent)
    .map((c) => c.text)
    .join("\n");
}

export function createUserMessage(
  text: string,
  attachments?: MessageAttachment[],
): Message {
  return {
    id: crypto.randomUUID(),
    role: "user",
    created: Date.now(),
    content: [{ type: "text", text }],
    metadata: attachments ? { attachments } : undefined,
  };
}

export function createSystemNotificationMessage(
  text: string,
  notificationType: SystemNotificationContent["notificationType"] = "info",
): Message {
  return {
    id: crypto.randomUUID(),
    role: "system",
    created: Date.now(),
    content: [{ type: "systemNotification", notificationType, text }],
    metadata: {
      userVisible: true,
      agentVisible: false,
    },
  };
}

File: ui/goose2/src-tauri/src/types/messages.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Message {
    pub id: String,
    pub role: MessageRole,
    pub created: i64,
    pub content: Vec<MessageContent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<MessageMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MessageRole {
    User,
    Assistant,
    System,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum MessageContent {
    Text {
        text: String,
    },
    Image {
        source: ImageSource,
    },
    ToolRequest {
        id: String,
        name: String,
        arguments: serde_json::Value,
        status: ToolCallStatus,
    },
    ToolResponse {
        id: String,
        name: String,
        result: String,
        #[serde(rename = "isError")]
        is_error: bool,
    },
    Thinking {
        text: String,
    },
    RedactedThinking {},
    Reasoning {
        text: String,
    },
    ActionRequired {
        id: String,
        #[serde(rename = "actionType")]
        action_type: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        message: Option<String>,
    },
    SystemNotification {
        #[serde(rename = "notificationType")]
        notification_type: String,
        text: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ImageSource {
    Base64 { media_type: String, data: String },
    Url { url: String },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ToolCallStatus {
    Pending,
    Executing,
    Completed,
    Error,
    Stopped,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct MessageMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_visible: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_visible: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attachments: Option<Vec<MessageAttachment>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub persona_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub persona_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_persona_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_persona_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completion_status: Option<MessageCompletionStatus>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum MessageCompletionStatus {
    InProgress,
    Completed,
    Error,
    Stopped,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageAttachment {
    #[serde(rename = "type")]
    pub attachment_type: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenState {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub total_tokens: u64,
    pub accumulated_input: u64,
    pub accumulated_output: u64,
    pub accumulated_total: u64,
}
